src/pdf.sandbox.js

```diff
@@ -23,87 +24,117 @@ const pdfjsBuild = PDFJSDev.eval("BUNDLE_BUILD");
   }
 
   create(data) {
-    const sandboxData = JSON.stringify(data);
-    const extra = [
-      "send",
-      "setTimeout",
-      "clearTimeout",
-      "setInterval",
-      "clearInterval",
-      "crackURL",
-    ];
-    const extraStr = extra.join(",");
-    let code = [
-      "exports = Object.create(null);",
-      "module = Object.create(null);",
+    const sandboxData = window.JSON.stringify(data);
+    const code = [
       // Next line is replaced by code from initialization.js
       // when we create the bundle for the sandbox.
       PDFJSDev.eval("PDF_SCRIPTING_JS_SOURCE"),
-      `data = ${sandboxData};`,
-      `module.exports.initSandbox({ data, extra: {${extraStr}}, out: this});`,
-      "delete exports;",
-      "delete module;",
-      "delete data;",
+      `pdfjsScripting.initSandbox({ data: ${sandboxData}, keepCallExternalFunction: ${TESTING} })`,
     ];
     if (!TESTING) {
-      code = code.concat(extra.map(name => `delete ${name};`));
-      code.push("delete debugMe;");
+      code.push("delete dump;");
+    } else {
+      code.unshift(
+        `globalThis.sendResultForTesting = callExternalFunction.bind(null, "send");`
+      );
+    }
+
+    let success = false;
+    try {
+      success = !!this._module.ccall(
+        "init",
+        "number",
+        ["string", "number"],
+        [code.join("\n"), this._alertOnError]
+      );
+    } catch (error) {
+      console.error(error);
+    }
+
+    if (success) {
+      this.support.commFun = this._module.cwrap("commFun", null, [
+        "string",
+        "string",
+      ]);
+    } else {
+      this.nukeSandbox();
     }
-    this._evalInSandbox(code.join("\n"), this._alertOnError);
-    this._dispatchEventName = data.dispatchEventName;
+
+    return success;
   }
 
   dispatchEvent(event) {
-    if (this._dispatchEventName === null) {
-      throw new Error("Sandbox must have been initialized");
-    }
-    event = JSON.stringify(event);
-    this._evalInSandbox(
-      `app["${this._dispatchEventName}"](${event});`,
-      this._alertOnError
-    );
+    this.support.callSandboxFunction("dispatchEvent", event);
   }
 
   dumpMemoryUse() {
-    this._module.ccall("dumpMemoryUse", null, []);
+    if (this._module) {
+      this._module.ccall("dumpMemoryUse", null, []);
+    }
   }
 
   nukeSandbox() {
-    this._dispatchEventName = null;
-    this._module.ccall("nukeSandbox", null, []);
-    this._module = null;
-    this._evalInSandbox = null;
+    if (this._module !== null) {
+      this.support.destroy();
+      this.support = null;
+      this._module.ccall("nukeSandbox", null, []);
+      this._module = null;
+    }
   }
 
   evalForTesting(code, key) {
     if (TESTING) {
-      this._evalInSandbox(
-        `try {
-           send({ id: "${key}", result: ${code} });
-         } catch (error) {
-           send({ id: "${key}", result: error.message });
-         }`,
-        this._alertOnError
+      this._evalInSandbox = this._module.ccall(
```

Comment:
Remove `this._evalInSandbox = `. You're using `ccall`, not `cwrap`, so there is no return value.